File: 0101-0200/0152.Maximum_Product_Subarray.js
//2ms





/**
 * @param {number[]} nums
 * @return {number}
 */


//  HAil Striver  ************
var maxProduct = function(nums)   {      
    let prefix =1
    let suffix = 1
    let ans=nums[0]

    for(let i=0; i<nums.length; i++){
        if(prefix===0){
            prefix=1
        }
        if(suffix===0){
            suffix=1
        }

        suffix= suffix*nums[i]
        prefix= prefix*nums[nums.length-i-1]
        ans=Math.max(ans,prefix,suffix)
    }
    return ans
};







//3ms





/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    let i = 0;
    let n = nums.length;
    let prefix = 1;
    let suffix = 1;
    let ans = nums[0];

    while(i<n){
        if(prefix == 0) prefix = 1;
        if(suffix == 0) suffix = 1;
        
        prefix *= nums[i];
        suffix *= nums[n-i-1];
        ans = Math.max(ans,suffix, prefix);
        i++;
    }

    return ans;
};

File: 0101-0200/0152.Maximum_Product_Subarray.java
//0ms





class Solution {
    public int maxProduct(int[] nums) {
        int Lproduct = 1;
        int Rproduct = 1;
        int max = nums[0];

        // Left to Right Pass
        for (int i = 0; i < nums.length; i++) {
            Rproduct *= nums[i];
            max = Math.max(max, Rproduct);
            if (Rproduct == 0) {
                Rproduct = 1;
            }
        }

        // Right to Left Pass
        for (int i = nums.length - 1; i >= 0; i--) {
            Lproduct *= nums[i];
            max = Math.max(max, Lproduct);
            if (Lproduct == 0) {
                Lproduct = 1;
            }
        }

        return max;
    }
}








//2ms








class Solution {
    public int maxProduct(int[] nums) {
        int max = Integer.MIN_VALUE;
        int prod = 1;
        for(int i=0;i<nums.length;i++){
            prod = prod*nums[i];
            max = Math.max(max,prod);
            if(prod==0)prod =1;
        }
        prod=1;
        for(int i=nums.length-1;i>=0;i--){
            prod = prod*nums[i];
            max = Math.max(max,prod);
            if(prod==0)prod =1;
        }
        return max;
    }
}
